Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -586,223 +499,90 @@ extension Decimal128: Hashable {
     }
 }
 
-/// Extension of `Decimal128` to add `BSONNumber` conformance.
-/// TODO: implement the missing converters (SWIFT-367)
-extension Decimal128 {
-    /// Create an `Int` from this `Decimal128`.
-    /// Note: this function is not implemented yet and will always return nil.
-    public var intValue: Int? { return nil }
-
-    /// Create an `Int32` from this `Decimal128`.
-    /// Note: this function is not implemented yet and will always return nil.
-    public var int32Value: Int32? { return nil }
-
-    /// Create an `Int64` from this `Decimal128`.
-    /// Note: this function is not implemented yet and will always return nil.
-    public var int64Value: Int64? { return nil }
-
-    /// Create a `Double` from this `Decimal128`.
-    /// Note: this function is not implemented yet and will always return nil.
-    public var doubleValue: Double? { return nil }
-
-    /// Returns this `Decimal128`.
-    /// This is implemented as part of `BSONNumber` conformance.
-    public var decimal128Value: Decimal128? { return self }
-}
-
 /// An extension of `Double` to represent the BSON Double type.
-extension Double: BSONNumber {
-    public var bsonType: BSONType { return .double }
+extension Double: BSONValue {
+    internal static var bsonType: BSONType { return .double }
+
+    internal var bson: BSON { return .double(self) }
 
-    public func encode(to storage: DocumentStorage, forKey key: String) throws {
+    internal func encode(to storage: DocumentStorage, forKey key: String) throws {
         guard bson_append_double(storage._bson, key, Int32(key.utf8.count), self) else {
             throw bsonTooLargeError(value: self, forKey: key)
         }
     }
 
-    public static func from(iterator iter: DocumentIterator) throws -> Double {
+    internal static func from(iterator iter: DocumentIterator) throws -> BSON {
         guard iter.currentType == .double else {
             throw wrongIterTypeError(iter, expected: Double.self)
         }
 
-        return iter.withBSONIterPointer { iterPtr in
+        return .double(iter.withBSONIterPointer { iterPtr in
             self.init(bson_iter_double(iterPtr))
-        }
-    }
-}
-
-/// An extension of `Int` to represent the BSON Int32 or Int64 type.
-/// On 64-bit systems, `Int` corresponds to a BSON Int64. On 32-bit systems, it corresponds to a BSON Int32.
-extension Int: BSONNumber {
```

Comment:
since `Int` doesn't correspond to a `BSON` case, it no longer is a `BSONValue`.

`Int` can still be used with our BSON library in the following ways:
- Encoding/Decoding (e.g. having an `Int` property of a struct)
- Integer literals (e.g. `let x: Document = ["x": 1]`)
- through an initializer on `BSON` (e.g. `let x: Document = ["x": BSON(someInt)]`)
- through a conversion accessor on `BSON` (e.g. `x["x"]?.asInt()`)

Users also always have the option of explicit conversion (e.g. `x["x"] = .int64(Int64(someInt))`)